test/regress_dns.c

```diff
@@ -2114,6 +2115,136 @@ test_getaddrinfo_async(void *arg)
+	    &hints, gai_cb, &b_out[1]);
+	if (!r) {
+			// check
+			tt_int_op(b_out[1].err, ==, 0);
+			tt_assert(b_out[1].ai);
+			tt_assert(! b_out[1].ai->ai_next);
+			test_ai_eq(b_out[1].ai, "18.52.86.120:8001", SOCK_STREAM, IPPROTO_TCP);
+			tt_assert(b_out[1].ai->ai_canonname == NULL);
+	}
+
+	/* 2: v6only.example.com should have been cached */
+	hints.ai_family = PF_INET6;
+	hints.ai_flags = 0;
+	r = evdns_getaddrinfo(dns_base, "v6only.example.com", "8002",
+	    &hints, gai_cb, &b_out[2]);
+	tt_assert(!r);
+	// check
+	tt_int_op(b_out[2].err, ==, 0);
+	tt_assert(b_out[2].ai);
+	tt_assert(! b_out[2].ai->ai_next);
+	test_ai_eq(b_out[2].ai, "[b0b::f00d]:8002", SOCK_STREAM, IPPROTO_TCP);
+
+	/* 2.5: v6only.example.com cache lookup with PF_INET should return NULL addressinfo. */
+	hints.ai_family = PF_INET;
+	hints.ai_flags = 0;
+	evutil_freeaddrinfo(b_out[2].ai); // since this is reused
+	++n_gai_results_pending;
+	r = evdns_getaddrinfo(dns_base, "v6only.example.com", "8002",
+	    &hints, gai_cb, &b_out[2]);
+	tt_assert(!r);
+	// check
+	tt_int_op(b_out[2].err, ==, 0);
+	tt_assert(! b_out[2].ai);
+
+	/* 3: v4assert.example.com should have been cached */
+	hints.ai_family = PF_INET;
+	r = evdns_getaddrinfo(dns_base, "v4assert.example.com", "8003",
+	    &hints, gai_cb, &b_out[3]);
+	tt_assert(!r);
+	// check
+	tt_int_op(b_out[3].err, ==, 0);
+	tt_assert(b_out[3].ai);
+	tt_assert(! b_out[3].ai->ai_next);
+	test_ai_eq(b_out[3].ai, "18.52.86.120:8003", SOCK_STREAM, IPPROTO_TCP);
+
+	/* 4: v6assert.example.com should have been cached. */
+	hints.ai_family = PF_INET6;
+	r = evdns_getaddrinfo(dns_base, "v6assert.example.com", "8004",
+	    &hints, gai_cb, &b_out[4]);
+	tt_assert(!r);
+	/* check */
+	tt_int_op(b_out[4].err, ==, 0);
+	tt_assert(b_out[4].ai);
+	tt_assert(! b_out[4].ai->ai_next);
+	test_ai_eq(b_out[4].ai, "[b0b::f00d]:8004", SOCK_STREAM, IPPROTO_TCP);
+
+	/* 5: NEXIST shouldn't be cached, as it is instant. */
+	hints.ai_family = PF_INET;
+	r = evdns_getaddrinfo(dns_base, "nosuchplace.example.com", "8005",
+	    &hints, gai_cb, &b_out[5]);
+	tt_assert(r);
+
+	/* 6: NEXIST shouldn't be cached. */
+	hints.ai_family = PF_UNSPEC;
+	r = evdns_getaddrinfo(dns_base, "nosuchplace.example.com", "8006",
+	    &hints, gai_cb, &b_out[6]);
+	tt_assert(r);
+
+	/* 7: v6timeout.example.com timed out and therefore shouldn't be in cache. */
+	hints.ai_family = PF_UNSPEC;
+	r = evdns_getaddrinfo(dns_base, "v6timeout.example.com", "8007",
+	    &hints, gai_cb, &b_out[7]);
+	tt_assert(r);
+
+	/* 8: v6timeout-nonexist.example.com produced NEXIST and shouldn't be cached */
+	hints.ai_family = PF_UNSPEC;
+	r = evdns_getaddrinfo(dns_base, "v6timeout-nonexist.example.com",
+	    "8008", &hints, gai_cb, &b_out[8]);
+	tt_assert(r);
+
+	/* 9: AI_ADDRCONFIG should at least not crash. */
+	hints.ai_flags |= EVUTIL_AI_ADDRCONFIG;
+	r = evdns_getaddrinfo(dns_base, "both.example.com",
+	    "8009", &hints, gai_cb, &b_out[9]);
```

Comment:
What is the caching behavior? I guess it should not be cached, let's add `tt_assert(r)` then.

Reply:
I assumed we just check that it doesn't crash? This wouldn't make much sense:

```
if(r) tt_assert(r)
else tt_assert(!r)
```

Reply:
yeah it probably shouldn't cache [ece91c2](https://github.com/libevent/libevent/pull/1717/commits/ece91c2f31d29e2b73d529dcfc547c2864d88c09)